File: kirun-js/src/engine/function/system/string/Reverse.ts
import { Schema } from '../../../json/schema/Schema';
import { Event } from '../../../model/Event';
import { EventResult } from '../../../model/EventResult';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { AbstractFunction } from '../../AbstractFunction';
import { MapUtil } from '../../../util/MapUtil';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { TypeUtil } from '../../../json/schema/type/TypeUtil';
import { SchemaType } from '../../../json/schema/type/SchemaType';

export class Reverse extends AbstractFunction {
    protected readonly VALUE: string = 'value';

    private readonly signature: FunctionSignature = new FunctionSignature('Reverse')
        .setNamespace(Namespaces.STRING)
        .setParameters(
            new Map([
                [
                    this.VALUE,
                    new Parameter(this.VALUE, Schema.ofString(this.VALUE)).setVariableArgument(
                        false,
                    ),
                ],
            ]),
        )
        .setEvents(
            new Map([
                Event.outputEventMapEntry(
                    new Map([
                        [
                            this.VALUE,
                            new Schema()
                                .setType(TypeUtil.of(SchemaType.STRING))
                                .setName(this.VALUE),
                        ],
                    ]),
                ),
            ]),
        );

    public constructor() {
        super();
    }

    private readonly signature = this.signature;
    public getSignature(): FunctionSignature {
        return this.signature;
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        let acutalString: string = context.getArguments()?.get(this.VALUE);
        let stringLength: number = acutalString.length - 1;
        let reversedString: string = '';

        while (stringLength >= 0) {
            reversedString += acutalString.charAt(stringLength--);
        }

        return new FunctionOutput([EventResult.outputOf(MapUtil.of(this.VALUE, reversedString))]);
    }
}

File: kirun-py/src/kirun_py/function/system/string/reverse.py
from __future__ import annotations

from typing import TYPE_CHECKING

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.json.schema.type.schema_type import SchemaType
from kirun_py.json.schema.type.type_util import TypeUtil
from kirun_py.model.event import Event
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.model.function_signature import FunctionSignature
from kirun_py.model.parameter import Parameter
from kirun_py.namespaces.namespaces import Namespaces

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters


class Reverse(AbstractFunction):

    VALUE: str = 'value'

    def __init__(self) -> None:
        super().__init__()

        self._signature = (
            FunctionSignature('Reverse')
            .set_namespace(Namespaces.STRING)
            .set_parameters({
                self.VALUE: Parameter(
                    self.VALUE,
                    Schema.of_string(self.VALUE),
                ).set_variable_argument(False),
            })
            .set_events(dict([
                Event.output_event_map_entry({
                    self.VALUE: Schema()
                    .set_type(TypeUtil.of(SchemaType.STRING))
                    .set_name(self.VALUE),
                })
            ]))
        )

    def get_signature(self) -> FunctionSignature:
        return self._signature

    async def internal_execute(
        self, context: FunctionExecutionParameters
    ) -> FunctionOutput:
        actual_string: str = context.get_arguments().get(self.VALUE)
        reversed_string = actual_string[::-1]

        return FunctionOutput([
            EventResult.output_of({self.VALUE: reversed_string})
        ])
